design/open/iceberg-branching.md

```diff
@@ -0,0 +1,88 @@
+# Iceberg support: branching & isolation
+
+## Problem Description
+
+Iceberg stores metadata files that represent a snapshot of a given Iceberg table.
+Those metadata files save the location of the data hard-coded.
+This is problematic with lakeFS, since we cannot work on the same files with different branches.
+
+For example when working on branch `main`, the metadata files will be written with the property `"location" : "s3a://example-repo/main/db/table1"`.
+This means the data is located in `"s3a://example-repo/main/db/table1"`.
+However, when branching out from main to `feature-branch`, the metadata files still point to the main branch.
+With this we lose the capability of isolation.
+
+
+## Goals
+
+Enable working in isolation with an Iceberg table. Create a branch in lakeFS, and operate on the table on the branch.
+At first stage: support iceberg without being catalog agnostic - meaning work with only lakeFS catalog and not be compatible with other catalogs.
+At second stage: being catalog agnostic. Users will be able to configure both lakeFS catalog and other catalog together.
+- We will start by supporting a limited set of catalogs, for example Glue, Snowflake and Tabular.
+
+## Non Goals
+
+- Support merge operation.
+- Enable lakeFS operations using a dedicated SQL syntax.
+
+
+## Proposed Design: Dedicated lakeFS catalog
+
+The catalog stores current metadata pointer for Iceberg tables.
+With lakeFS, this pointer needs to be versioned as well.
+In order to prevent from adding versioning capabilities to existing catalogs, we'll extend a catalog that can utilize lakeFS to keep the pointer.
+Therefore, we'll extend the Hadoop catalog to work with lakeFS
+- Every operation of the catalog happens with the scope of a reference in lakeFS. This will be extracted from the table path.
+- Write location on metadata files to be relative: the repo and brach will not be written.
+- The catalog will know to use the metadata files from the relevant references
+  
+Implement LakeFSFileIO:
+[FileIO](https://iceberg.apache.org/javadoc/master/org/apache/iceberg/io/FileIO.html) is the primary interface between the core Iceberg library and underlying storage (read more [here](https://tabular.io/blog/iceberg-fileio/#:~:text=FileIO%20is%20the%20primary%20interface,how%20straightforward%20the%20interface%20is.)). In practice, it is just two operations: open for read and open for write.
+We'll Extend hadoop fileIO to work with lakeFS, and to be compatible with lakeFS catalog.
+As with the catalog, the reference in lakeFS will be extracted from the table path.
+
+With this design the catalog won't support atomicity. We believe it's Ok at first stage since we are always at scope of a branch.
+For that, we'll require a single-writer per branch.
+
```

Comment:
```suggestion
This design will require users to only have a single writer per branch. This because the Hadoop catalog uses a rename operation which is not atomic in lakeFS. We think it's a reasonable limitation to start with. We may be able to overcome this limitation in the future, using the `IfNoneMatch` flag in the lakeFS API.
```